File: src/main/java/gg/mineral/practice/listeners/EntryListener.java
package gg.mineral.practice.listeners;

import gg.mineral.bot.api.BotAPI;
import gg.mineral.practice.entity.PlayerStatus;
import gg.mineral.practice.managers.EloManager;
import gg.mineral.practice.managers.ProfileManager;
import gg.mineral.practice.scoreboard.impl.DefaultScoreboard;
import lombok.val;
import org.bukkit.GameMode;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerInitialSpawnEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;

public class EntryListener implements Listener {

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        event.setJoinMessage(null);
        ProfileManager.removeIfExists(event.getPlayer());
        val profile = ProfileManager.getOrCreateProfile(event.getPlayer());
        profile.setGameMode(GameMode.SURVIVAL);
        profile.heal();

        if (BotAPI.INSTANCE.isFakePlayer(profile.getPlayer().getUniqueId()))
            return;

        EloManager.updateName(profile);
        profile.getInventory().setInventoryForLobby();
        profile.removePotionEffects();
        profile.setPlayerStatus(PlayerStatus.IDLE);
        profile.setScoreboard(DefaultScoreboard.INSTANCE);
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent e) {
        e.setQuitMessage(null);

        val victim = ProfileManager.getOrCreateProfile(e.getPlayer());

        victim.removeScoreboard();

        if (victim.isInParty())
            victim.getParty().leave(victim);
        else if (victim.isInTournament())
            victim.getTournament().removePlayer(victim);
        else if (victim.isInEvent())
            victim.getEvent().removePlayer(victim);

        switch (victim.getPlayerStatus()) {
            case FIGHTING -> victim.getMatch().end(victim);
            case QUEUEING -> victim.removeFromQueue();
            default -> {
            }
        }

        ProfileManager.remove(victim);
    }

    @EventHandler
    public void onPlayerInitialSpawn(PlayerInitialSpawnEvent e) {

        if (BotAPI.INSTANCE.isFakePlayer(e.getPlayer().getUniqueId()))
            return;

        e.setSpawnLocation(ProfileManager.getSpawnLocation());
    }
}

File: src/main/kotlin/gg/mineral/practice/listeners/EntryListener.kt
package gg.mineral.practice.listeners

import gg.mineral.practice.entity.PlayerStatus
import gg.mineral.practice.entity.appender.PlayerAppender
import gg.mineral.practice.managers.EloManager.updateName
import gg.mineral.practice.managers.ProfileManager.getOrCreateProfile
import gg.mineral.practice.managers.ProfileManager.lobbyLocation
import gg.mineral.practice.managers.ProfileManager.remove
import gg.mineral.practice.managers.ProfileManager.removeIfExists
import gg.mineral.practice.scoreboard.impl.DefaultScoreboard
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerInitialSpawnEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent

class EntryListener : Listener, PlayerAppender {

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        event.joinMessage = null
        removeIfExists(event.player)
        event.player.gameMode = GameMode.SURVIVAL
        event.player.heal()

        if (event.player.isFake()) return

        event.player.removePotionEffects()
        updateName(event.player)

        val profile = getOrCreateProfile(event.player)
        profile.inventory.setInventoryForLobby()

        profile.playerStatus = PlayerStatus.IDLE
        profile.scoreboard = DefaultScoreboard.INSTANCE
    }

    @EventHandler
    fun onPlayerQuit(e: PlayerQuitEvent) {
        e.quitMessage = null

        val victim = getOrCreateProfile(e.player)

        victim.taskIds.forEach { Bukkit.getScheduler().cancelTask(it) }

        victim.party?.leave(victim)
        victim.contest?.removePlayer(victim)

        when (victim.playerStatus) {
            PlayerStatus.FIGHTING -> victim.match?.end(victim)
            PlayerStatus.QUEUEING -> victim.removeFromQueue()
            else -> {}
        }

        remove(victim)
    }

    @EventHandler
    fun onPlayerInitialSpawn(e: PlayerInitialSpawnEvent) {
        if (e.player.isFake()) return
        e.spawnLocation = lobbyLocation
    }
}
